pkg/table-rule-selector/trie_selector.go

```diff
@@ -0,0 +1,459 @@
+}
+
+// RuleSet is a set of rules that selected
+type RuleSet []interface{}
+
+func (r RuleSet) clone() RuleSet {
+	if r == nil {
+		return nil
+	}
+
+	c := make(RuleSet, 0, len(r))
+	return append(c, r...)
+}
+
+type macthedResult struct {
+	nodes []*node
+	rules RuleSet
+}
+
+func (r *macthedResult) empty() bool {
+	return r == nil || (len(r.nodes) == 0 && len(r.rules) == 0)
+}
+
+type trieSelector struct {
+	sync.RWMutex
+
+	cache map[string]RuleSet
+	root  *node
+}
+
+type node struct {
+	characters         map[byte]*item
+	asterisk, question *item
+}
+
+type item struct {
+	child *node
+
+	rule interface{}
+	// schema level ->(to) table level
+	nextLevel *node
+}
+
+func newNode() *node {
+	return &node{characters: make(map[byte]*item)}
+}
+
+// NewTrieSelector returns a trie Selector
+func NewTrieSelector() Selector {
+	return &trieSelector{cache: make(map[string]RuleSet), root: newNode()}
+}
+
+// Insert implements Selector's interface.
+func (t *trieSelector) Insert(schema, table string, rule interface{}, replace bool) error {
+	if len(schema) == 0 || rule == nil {
+		return errors.Errorf("schema pattern %s or rule %v can't be empty", schema, rule)
+	}
+
+	var err error
+	t.Lock()
+	if len(table) == 0 {
+		err = t.insertSchema(schema, rule, replace)
+	} else {
+		err = t.InsertTable(schema, table, rule, replace)
+	}
+	t.Unlock()
+
+	return errors.Trace(err)
+}
+
+func (t *trieSelector) insertSchema(schema string, rule interface{}, replace bool) error {
+	_, err := t.insert(t.root, schema, rule, replace)
+	if err != nil {
+		return errors.Annotate(err, "insert into schema selector")
+	}
+
+	return nil
+}
+
+func (t *trieSelector) InsertTable(schema, table string, rule interface{}, replace bool) error {
+	schemaEntity, err := t.insert(t.root, schema, nil, false)
+	if err != nil {
+		return errors.Annotate(err, "insert into schema selector")
+	}
+
+	if schemaEntity.nextLevel == nil {
+		schemaEntity.nextLevel = newNode()
+	}
+
+	_, err = t.insert(schemaEntity.nextLevel, table, rule, replace)
+	if err != nil {
+		return errors.Annotate(err, "insert into table selector")
+	}
+
+	return nil
+}
+
+// if rule is nil, just extract nodes
+func (t *trieSelector) insert(root *node, pattern string, rule interface{}, replace bool) (*item, error) {
+	var (
+		n           = root
+		hadAsterisk = false
+		entity      *item
+	)
+
+	for i := range pattern {
+		if hadAsterisk {
+			return nil, errors.Errorf("pattern %s is invaild", pattern)
+		}
+
+		switch pattern[i] {
+		case asterisk:
+			entity = n.asterisk
+			hadAsterisk = true
+		case question:
+			entity = n.question
+		default:
+			entity = n.characters[pattern[i]]
+		}
+		if entity == nil {
+			entity = &item{}
+			switch pattern[i] {
+			case asterisk:
+				n.asterisk = entity
+			case question:
+				n.question = entity
+			default:
+				n.characters[pattern[i]] = entity
+			}
+		}
+		if entity.child == nil {
+			entity.child = newNode()
+		}
+		n = entity.child
+	}
+
+	if rule != nil {
```

Comment:
`rule != nil` is always `true` here.

as 1st statement in `Insert` method is as follows:

```
if len(schema) == 0 || rule == nil {
    return errors...
}
```

Reply:
```
schemaEntity, err := t.insert(t.root, schema, nil, false)
if err != nil {
     return errors.Annotate(err, "insert into schema selector")
}
```